protocol/consistencychecks.go

```diff
@@ -225,43 +229,65 @@ func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
 	// verify hash chain
 	// TODO: clearly verify which is the client's actual expectation.
 	// See #81
-	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
-		(str.Epoch == cc.CurrentEpoch+1 && str.VerifyHashChain(cc.SavedSTR)) {
+	if (str.Epoch == cc.SavedSTR.Epoch && bytes.Equal(cc.SavedSTR.Signature, str.Signature)) ||
+		(str.Epoch == cc.SavedSTR.Epoch+1 && str.VerifyHashChain(cc.SavedSTR.Signature)) {
 		return nil
 	}
+
+	// TODO: verify the directory's policies as well. See #115
 	return ErrorBadSTR
 }
 
-// verifySignedPromise checks if the returned response is match
-// with the directory's policy. Then it calls verifyFulfilledPromise
-// to verify whether issued TBs was inserted in the directory
-// as promised. Finally, it verifies the returned promise
-// based on the request type. Note that the directory returns a promise
-// iff the returned proof is a proof of absence.
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	for name, tb := range cc.TBs {
+		if cc.SavedSTR.Epoch+1 == str.Epoch &&
+			name == uname {
+			if !tb.Verify(ap) {
+				return ErrorBrokenPromise
+			}
+			delete(cc.TBs, uname)
+			return nil
+		} else if cc.SavedSTR.Epoch < str.Epoch {
+			// TODO: should verify when we introduce KeyLookupInEpoch
+			// and monitoring with missed epochs,
+			// see: https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84930999
+			// and: https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84501211
+			delete(cc.TBs, name)
+		}
+		// keep current epoch's returned promises
+		// for future verifications
+	}
+	return nil
```

Comment:
> I think the point is the TBs should be verified using `cc.SavedSTR`, it means the str in (tb, str) is already stored in `cc.SavedSTR`, right?

You're right, any TBs should only ever be verified against `cc.SavedSTR` assuming that `ccSavedSTR.Epoch == tb_epoch +1`, i.e. the client has verified the next STR. But if we're not clearing the TBs at the beginning of a new epoch, the client cannot distinguish between old TBs that should have been verified in a previous epoch and recent TBs that correspond to `cc.SavedSTR`. 

In other words, I think the only way that the client can know whether it expects a TB for a lookup in the current implementation is if it has a notion of epoch transitions, i.e. the client does a monitoring operation on a regular basis (when the epoch deadline in the current `cc.SavedSTR.Policy` is reached). Then it would be safe for it to clear any TBs that it still has cached at that point until we address #116. What do you think?